src/main/java/net/sf/jabref/gui/JabRefFrame.java

```diff
@@ -1927,16 +1928,22 @@ private int addBibEntries(List<BibtexEntry> bibentries, String filename,
                         if (res == DuplicateResolverDialog.KEEP_LOWER) {
                             dupli = true;
                         } else if (res == DuplicateResolverDialog.KEEP_UPPER) {
-                            database.removeEntry(existingEntry.getId());
-                            ce.addEdit(new UndoableRemoveEntry
-                                    (database, existingEntry, basePanel));
+                            remove = true;
+                        } else if (res == DuplicateResolverDialog.KEEP_MERGE) {
+                            LOGGER.error("Merge button!");
+                            dupli = true;
+                            remove = true;
+                            toAdd.add(drd.getMergedEntry());
                         } else if (res == DuplicateResolverDialog.BREAK) {
                             break mainLoop;
                         }
+                        if (remove) {
+                            database.removeEntry(existingEntry.getId());
+                            ce.addEdit(new UndoableRemoveEntry(database, existingEntry, basePanel));
```

Comment:
why are entries removed directly from the DB, but the addition of entries is delayed until after the loop?